pkg/client/simple/clientset.go

```diff
@@ -53,10 +53,10 @@ type Clientset interface {
 	SecretStore(cluster *kops.Cluster) (fi.SecretStore, error)
 
 	// KeyStore builds the key store for the specified cluster
-	KeyStore(cluster *kops.Cluster) (fi.CAStore, error)
+	KeyStore(ctx context.Context, cluster *kops.Cluster) (fi.CAStore, error)
```

Comment:
This shouldn't take a Context

Reply:
Not sure I follow.  I think all our methods are going to end up taking a context.  And this one directly needs a context for VFSClientset::KeyStore -> pkiPath -> VFSContext::BuildVfsPath

Reply:
The methods on a CAStore might take a context, but it doesn't make much sense for the CAStore itself to hold a context.

I'm thinking a lot of the problem has to do with `vfs.Path` implementing `io.WriterTo`. Perhaps one should call a method taking a context in order to get a `io.WriterTo`.

Reply:
I think that's a valid concern, and I've been wondering if it's possible to change where we _need_ the context.  The reason it's baked into the VFSContext right now is that we are caching a GCS client, which includes the http.Client, which is what we're going to intercept/override in our tests.  Arguably the GCS Client and the http.Client should be built on every method call, but I _think_ the GCS client is not just using the standard http.Transport.  So maybe we should be caching it ... etc etc

But: I don't think we should assume that a method that takes a context.Context is going to embed the context.Context.  If the context means anything, it means that the method is non-trivial.  Anything that logs, anything that emits an otel span, or anything that calls any method that does that is going to take a context parameter.

So, while I think we can absolutely look at where we store the GCS client and friends, overall we're going to have a context on basically every exported function, and probably most internal functions also.

I suspect we'll do this for a few years until we decide that maybe goroutine-local-storage is a good idea, and then we'll hopefully just have a bulk refactoring tool to undo all the context args ;-)

Reply:
The design pattern I'm applying is that contexts don't belong in clientsets. One should mock out the client through the context; one should mock out the client through the clientset.

The VFSContext (which is a clientset) should not be a global variable; it should be passed in explicitly. And `vfs.Path` is not just a path but something which currently inherently needs to embed a context.

Reply:
OK, that makes sense and I will look at doing that.  I think we need to pass it into every callsite to BuildVfsPath (which can then pass a context to buildGCSPath).  I think the GCS vfs.Path will end up embedding a GCS client, which I think is equivalent.

But, in order to get there, we need to wire up the context into a lot of methods.  I can either do that piecemeal in "waves" - one idea or set of functions at a time - with the hope that these smaller PRs can be more easily reviewed, or I can do a "big bang" where we change every callsite.  I feel that changing every callsite at once will be almost impossible to review though...

This particular function calls VFSClientset::KeyStore -> pkiPath -> VFSContext::BuildVfsPath anyway, so we'd need the context argument even if I did the "BuildVfsPath" wave immediately.  (And that "wave" looks like it'll be _very_ big)

Reply:
(The PR where I tried doing _some_ of this is https://github.com/kubernetes/kops/pull/14777 and it rapidly becomes all-encompassing...)

Reply:
I don't think we should hang a `VFSContext` off of a `context.Context` and I don't think we should be adding `context.Context` parameters to enable hanging a `VFSContext` off of a `context.Context`.

It would be reasonable for a `VFSClientset` to have a `VFSContext`.

Reply:
OK, I changed this PR to _only_ wire up context.Context into more functions.

Once we have context.Context available in enough callsites, we can try it the more logically coherent way.